docs/src/proposals/page-indexed-transactions.md

```diff
@@ -0,0 +1,235 @@
+ensure fast and reliable network transmission of cluster info over UDP.
+Solana's networking stack uses a conservative MTU size of 1280 bytes which,
+after accounting for headers, leaves 1232 bytes for packet data like serialized
+transactions.
+
+Developers building applications on Solana must design their on-chain program
+interfaces within the above transaction size limit constraint. One common
+work-around is to store state temporarily on-chain and consume that state in
+later transactions. This is the approach used by the BPF loader program for
+deploying Solana programs.
+
+However, this workaround doesn't work well when developers compose many on-chain
+programs in a single atomic transaction. With more composition comes more
+account inputs, each of which takes up 32 bytes. There is currently no available
+workaround for increasing the number of accounts used in a single transaction
+since each transaction must list all accounts that it needs to properly lock
+accounts for parallel execution. Therefore the current cap is about 35 accounts
+after accounting for signatures and other transaction metadata.
+
+## Proposed Solution
+
+Introduce a new on-chain account indexing program which stores account address
+mappings and add a new transaction format which supports concise account
+references through on-chain account indexes.
+
+### Account Indexing Program
+
+Here we describe a contract-based solution to the problem, whereby a protocol
+developer or end-user can create collections of related accounts on-chain for
+concise use in a transaction's account inputs. This approach is similar to page
+tables used in operating systems to succinctly map virtual addresses to physical
+memory.
+
+After addresses are stored on-chain in an index account, they may be succinctly
+referenced from a transaction using an index rather than a full 32 byte address.
+This will require a new transaction format to make use of these succinct indexes
+as well as runtime handling for looking up and loading accounts from the
+on-chain indexes.
+
+#### State
+
+Index accounts must be rent-exempt and may not be deleted. Stored addresses
+should be append only so that once an address is stored in an index account, it
+may not be removed later. Index accounts must be pre-allocated since Solana does
+not support reallocation for accounts yet.
+
+Since transactions use a u16 offset to look up addresses, index accounts can
+store up to 2^16 addresses each. Anyone may create an index account of any size
+as long as its big enough to store the necessary metadata. In addition to
+stored addresses, index accounts must also track the latest count of stored
+addresses and an authority which must be a present signer for all index
+additions.
+
+#### Cost
+
+Since index accounts require caching and special handling in the runtime, they should incur
+higher costs for storage. Cost structure design will be added later.
```

Comment:
Implement this on devnet/testnet first, without updates to the cost model to account for this, but don't activate on mainnet?  Just trying to establish what would trigger "later" in the future.

Reply:
Should be before devnet/testnet. I felt that the cost structure could be hashed out as part of the ongoing cost model work rather than within the context of this proposal. But it is important to figure out so how about I kick off the discussion with a few rough ideas..

First, the main concern here is having many large indexes that all need to be cached in memory due to high usage. We need some way to cap memory usage and fairly price-in the increase of RAM requirements.

Ideas:

1) Each newly initialized index should have higher storage cost than a typical Solana account. This could be some multiple of the cost of standard rent which operates similar to rent-exemption and allows accounts to stay in RAM

2) Charge fees on index accounts whenever they are used in transactions. Once the index reserve is depleted, the index is deactivated and causes transactions to fail until more funds are added.

Hey @taozhu-chicago want to chime in here?  Context is above, would love to hear your thoughts

Reply:
The indexes would currently be memory mapped like all other accounts by default?  I feel like we need a more general model of evicting long-unused accounts from RAM, with probably a new instruction to warm an account back up again once it's moved off to cold storage.  Frequently used programs and other large accounts would have the same issue really.

But yep, removing the cost model design from here makes sense to me.  Just want to ensure we don't forget about it somehow!

Reply:
Awesome, sounds good. Hoping this issue gets solved in a general way like that